hpx/parallel/segmented_algorithms/detail/reduce.hpp

```diff
@@ -0,0 +1,125 @@
+        {
+            return util::partitioner<ExPolicy, T>::call(
+                std::forward<ExPolicy>(policy),
+                first, std::distance(first, last),
+                [r](FwdIter part_begin, std::size_t part_size) -> T
+                {
+                    T val = *part_begin;
+                    return util::accumulate_n(++part_begin, --part_size,
+                        std::move(val), r);
+                },
+                hpx::util::unwrapped([r](std::vector<T> && results)
+                {
+                    auto rfirst = std::begin(results);
+                    auto rlast = std::end(results);
+                    return util::accumulate<T>(rfirst, rlast, r);
+                }));
+        }
+    };
+
+    template <typename T>
+    struct seg_transform_reduce
+      : public detail::algorithm<seg_transform_reduce<T>, T>
+    {
+        seg_transform_reduce()
+          : seg_transform_reduce::algorithm("transform_reduce")
+        {}
+
+        template <typename ExPolicy, typename InIter, typename Reduce,
+            typename Convert>
+        static T
+        sequential(ExPolicy, InIter first, InIter last, Reduce && r,
+            Convert && conv)
+        {
+            typedef typename std::iterator_traits<InIter>::reference
+                reference;
+            return util::accumulate<T>(first, last,
+                [=](T const& res, reference next)
+                {
+                    return hpx::util::invoke(r, res,
+                        hpx::util::invoke(conv, next));
+                },
+                std::forward<Convert>(conv)
+            );
+        }
+
+        template <typename ExPolicy, typename FwdIter, typename Reduce,
+            typename Convert>
+        static typename util::detail::algorithm_result<ExPolicy, T>::type
+        parallel(ExPolicy && policy, FwdIter first, FwdIter last, Reduce && r,
+            Convert && conv)
+        {
+            typedef typename std::iterator_traits<FwdIter>::reference
+                reference;
+
+            return util::partitioner<ExPolicy, T>::call(
+                std::forward<ExPolicy>(policy),
+                first, std::distance(first, last),
+                [r, conv](FwdIter part_begin, std::size_t part_size) -> T
+                {
+                    T val = hpx::util::invoke(conv, *part_begin);
+                    return util::accumulate_n(++part_begin, --part_size,
+                        std::move(val),
+                        // MSVC14 bails out if r and conv are captured by
+                        // reference
+                        [=](T const& res, reference next)
+                        {
+                            return hpx::util::invoke(r, res,
+                                hpx::util::invoke(conv, next));
+                        });
+                },
+                hpx::util::unwrapped([r](std::vector<T> && results)
+                {
+                    auto rfirst = std::begin(results);
```

Comment:
Please use `hpx::util::begin`/`hpx::util::end` instead of the standard versions.